### PR TITLE
rpc: Our rpcClient should make an attempt to reconnect.

### LINKED ISSUE #3205: [distributed] replace disk, heal disk does not work the first time
## Steps to Reproduce (for bugs)
* start 2:2 dXL
* bring down 4th server
* replace disk of 4th server (rm -rf and mkdir)
* bring up 4th server
* heal the disk:
```
krishna@escape:~$ minio control heal http://localhost:9001/
krishna@escape:~$ ls /export/disk4/ -a                             <---- disk not healed
.  ..
krishna@escape:~$ minio control heal http://localhost:9001/
krishna@escape:~$ ls /export/disk4/ -a
.  ..  .minio.sys                                                                    <- disk healed on 2nd try
krishna@escape:~$ 
```

i.e `minio control heal http://localhost:9001/` works only on the 2nd time.

## Expected Behavior

`minio control heal http://localhost:9001/` should work the first time.

## Current Behavior

`minio control heal http://localhost:9001/` works only on the 2nd time.

### LINKED ISSUE #3206: [distributed] replace disk, bucket heal does not happen in one situation
## Steps to Reproduce (for bugs)
* start 2:2 dXL
* bring down 4th server
* replace disk of 4th server (rm -rf and mkdir)
* bring up 4th server
* heal the disk:
```
krishna@escape:~$ minio control heal http://localhost:9004/
krishna@escape:~$ ls /export/disk4/ -a
.  ..  .minio.sys
krishna@escape:~$ 
```
* heal the bucket:
```
krishna@escape:~$ minio control heal http://localhost:9004/test

krishna@escape:~$ ls /export/disk4/ -a
.  ..  .minio.sys
krishna@escape:~$
```

Bucket is not healed. But it heals if healing is pointed towards localhost:9001/9002/9003

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
rpcClient should attempt a reconnect if the call fails
with 'rpc.ErrShutdown' this is needed since at times when
the servers are taken down and brought back up.

The hijacked connection from net.Dial is usually closed.

So upon first attempt rpcClient might falsely indicate that
disk to be down, to avoid this state make another dial attempt
to really fail.

## Motivation and Context
Fixes #3206
Fixes #3205

## How Has This Been Tested?
Manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/format-config-v1.go [hunk 1]
```diff
@@ -230,7 +230,8 @@ func genericFormatCheck(formatConfigs []*formatConfigV1, sErrs []error) (err err
 	)
 
 	// Through all errors calculate the actual errors.
-	for _, lErr := range sErrs {
+	for index, lErr := range sErrs {
+		fmt.Println(lErr, index)
```
Comment (vadmeste) [inline cmd/format-config-v1.go]: This is not needed.

File: cmd/net-rpc-client.go [hunk 2]
```diff
@@ -150,8 +150,18 @@ func (rpcClient *RPCClient) Call(serviceMethod string, args interface{}, reply i
...
+			rpcLocalStack, err = rpcClient.dialRPCClient()
+			if err != nil {
+				return err
+			}
+
+			// Attempt the service call oncne again, upon any error now just give up.
+			err = rpcLocalStack.Call(serviceMethod, args, reply)
+			if err != nil && err.Error() == rpc.ErrShutdown.Error() {
+				// Set rpc error as rpc.ErrShutdown type.
+				err = rpc.ErrShutdown
```
Comment (vadmeste) [inline cmd/net-rpc-client.go]: Isn't it useless to set err here ? because we enter here only when err is already rpc.ErrShutdown

Comment (harshavardhana) [reply cmd/net-rpc-client.go]: no we need to set it back to a proper type so that the type is not lost. otherwise rpc.Call() converts this into rpc.ServerError() at times not all the time.

File: cmd/format-config-v1.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/format-config-v1.go]: Oops yeah.

File: cmd/net-rpc-client.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/net-rpc-client.go]: Let me try again locally and see if the type is not lost in these situations if the type is not lost then we can keep it rpc.ErrShudown.

Comment (harshavardhana) [reply cmd/net-rpc-client.go]: what happened @donatello don't see your comment anymore?

Comment (harshavardhana) [reply cmd/net-rpc-client.go]: Just tried the type is lost. it doesn't work.

File: cmd/net-rpc-client.go [hunk 3]
```diff
@@ -141,20 +141,26 @@ func (rpcClient *RPCClient) Call(serviceMethod string, args interface{}, reply i
...
 	err := rpcLocalStack.Call(serviceMethod, args, reply)
-	if err != nil {
-		if err.Error() == rpc.ErrShutdown.Error() {
-			// Reset rpcClient.rpc to nil to trigger a reconnect in future
-			// and close the underlying connection.
-			rpcClient.clearRPCClient()
+	// Any errors other than rpc.ErrShutdown just return quickly.
+	if err != nil && err != rpc.ErrShutdown {
+		return err
+	} // else rpc.ErrShutdown returned by rpc.Call
```
Comment (donatello) [inline cmd/net-rpc-client.go]: doesnt seem to be right. It can be that `err == nil` after the if block..

Comment (harshavardhana) [reply cmd/net-rpc-client.go]: yes fixing this.

**Review by donatello — APPROVED** [review]
